connector/src/main/java/com/google/cloud/hive/bigquery/connector/BigQueryMetaHook.java

```diff
@@ -129,39 +147,177 @@ public void preCreateTable(Table table) throws MetaException {
+      // write method. In this case, we will stream-write to a temporary table
+      // and then finally overwrite the final destination table with the temporary
+      // table's contents. This special case doesn't apply to the 'indirect'
+      // write method, which doesn't need a temporary table -- instead that method
+      // uses the 'WRITE_TRUNCATE' option available in the BigQuery Load Job API when
+      // loading the Avro files into the BigQuery table (see more about that in the
+      // `IndirectOutputCommitter` class).
+      if (overwrite) {
+        // Set the final destination table as the job's original table
+        jobInfo.setFinalTable(tableName);
+        // Create a temporary table with the same schema
+        // TODO: It'd be useful to add a description to the table explaining that it was
+        //  created as a temporary table for a Hive query.
+        TableInfo tableInfo =
+            bqClient.createTempTable(
+                TableId.of(
+                    jobInfo.getProject(),
+                    jobInfo.getDataset(),
+                    tableName + "-" + HiveUtils.getHiveId(conf) + "-"),
+                bigQuerySchema);
+        // Set the temp table as the job's output table
+        jobInfo.setTable(tableInfo.getTableId().getTable());
+      }
+    } else if (writeMethod.equals(HiveBigQueryConfig.WRITE_METHOD_INDIRECT)) {
+      String tempGcsPath = conf.get(HiveBigQueryConfig.TEMP_GCS_PATH_KEY);
+      jobInfo.setGcsTempPath(tempGcsPath);
+      if (tempGcsPath == null || tempGcsPath.trim().equals("")) {
+        throw new MetaException(
+            String.format(
+                "The '%s' property must be set when using the '%s' write method.",
+                HiveBigQueryConfig.TEMP_GCS_PATH_KEY, HiveBigQueryConfig.WRITE_METHOD_INDIRECT));
+      } else if (!IndirectUtils.hasGcsWriteAccess(tempGcsPath)) {
+        throw new MetaException(
+            String.format(
+                "Cannot write to table '%s'. Does not have write access to the"
+                    + " following GCS path, or bucket does not exist: %s",
+                table.getTableName(), tempGcsPath));
+      }
+    } else {
+      throw new MetaException("Invalid write method: " + writeMethod);
+    }
+
+    // Save the info file so that we can retrieve all the information at later
+    // stages of the job's execution
+    JobInfo.writeInfoFile(conf, jobInfo);
+  }
 
+  /**
+   * This method is called automatically at the end of a successful job when using the "tez"
+   * execution engine. This method is not called when using "mr" -- for that, see {@link
+   * BigQueryOutputCommitter#commitJob(JobContext)}
+   */
   @Override
-  public void rollbackInsertTable(Table table, boolean overwrite) throws MetaException {}
+  public void commitInsertTable(Table table, boolean overwrite) throws MetaException {
+    String engine = HiveConf.getVar(conf, HiveConf.ConfVars.HIVE_EXECUTION_ENGINE);
+    if (engine.equals("tez")) {
```

Comment:
Yes it can, but that goes through a different workflow in `BigQueryOutputCommitter#commitJob(JobContext)`. This is mentioned in the method comment above.